### PR TITLE
Add macro handling and templating

### PR DESCRIPTION
**Pull request by Mierunski**

### CODE REVIEW

File: templates/default.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+STM32L496G_BG96_binary:
+  0. Download the `Anjay-freertos-client-STM32L496G-BG96.bin` file.
+L462_binary:
+  0. Download the `Anjay-freertos-client-B-L462E-CELL1-TYPE1SC.bin` file.
+zephyr_files:
+  0. Go to [Anjay-zephyr-client](https://github.com/AVSystem/Anjay-zephyr-client/releases).
+nRF9160_binary:
+  0. Download the `demo_nrf9160dk_merged.hex` file.
+thingy91_binary:
+  0. Download the `demo_thingy91_app_signed.hex` file.
```
Comment (JZimnol) [inline templates/default.yaml:16]:
```suggestion
coiote_long_name: Coiote IoT Device Management platform
coiote_site_link: https://eu.iot.avsystem.cloud
coiote_short_name: Coiote IoT DM
coaps_uri: coaps://eu.iot.avsystem.cloud:5684
freertos_repository_step:
  0. Go to [Anjay-freertos-client](https://github.com/AVSystem/Anjay-freertos-client/releases/).
STM32L496G_BG96_binary_step:
  0. Download the `Anjay-freertos-client-STM32L496G-BG96.bin` file.
B_L462_CELL1_binary:
  0. Download the `Anjay-freertos-client-B-L462E-CELL1-TYPE1SC.bin` file.
zephyr_repository_step:
  0. Go to [Anjay-zephyr-client](https://github.com/AVSystem/Anjay-zephyr-client/releases).
nRF9160DK_binary_step:
  0. Download the `demo_nrf9160dk_merged.hex` file.
Thingy91_binary_step:
  0. Download the `demo_thingy91_app_signed.hex` file.
```

**Review by JZimnol — COMMENTED** [review]
@Mierunski added suggestions about macro names

File: templates/default.yaml (hunk 1 above)
Comment (Mierunski) [reply templates/default.yaml:16]: done

File: docs/LwM2M_Client/Nordic/Thingy91.md [hunk 2]
```diff
@@ -132,9 +132,19 @@ To connect the board:
...
+
+ LwM2M Server URI: ''' + coaps_uri + '''
+ Device lifetime: 50
+ Endpoint name: test_device
+ PSK: psk
+ Bootstrap: n
+''')}}
 
-0. Update your device credentials:
+1. Update your device credentials:
```
Comment (JZimnol) [inline docs/LwM2M_Client/Nordic/Thingy91.md:147]: Why did you change numbering from `0.` to `1. 2. 3. ...` etc?

File: docs/LwM2M_Client/Nordic/Thingy91.md [hunk 3]
```diff
@@ -153,13 +163,13 @@ To connect the board:
...
 
-0. Start the client using the new configurations:
+2. Start the client using the new configurations:
 
     ```
     anjay start
     ```
 
-0. Go to the Coiote DM to check if your device is connected.
+3. Go to the Coiote DM to check if your device is connected.
```
Comment (JZimnol) [inline docs/LwM2M_Client/Nordic/Thingy91.md:172]:
```suggestion
3. Go to the {{ coiote_short_name }} to check if your device is connected.
```

File: docs/LwM2M_Client/ESP32/ESP32-DevKitCBG96.md [hunk 4]
```diff
@@ -19,7 +19,7 @@ Integrate your ESP32-based device with Quectel BG96 module to manage it via Coio
       ![BG96 UART example configuration](images/BG96_uart_config.png "BG96 UART example configuration")
     - In **Client options**, provide device credentials and Server URI:
         - **Endpoint name** - your device endpoint name
-        - **Server URI** - the address and port of your Coiote DM installation, e.g. ``coaps://eu.iot.avsystem.cloud:5684``
+        - **Server URI** - the address and port of your {{ coiote_short_name }} installation, e.g. ``coaps_uri://eu.iot.avsystem.cloud:5684``
```
Comment (JZimnol) [inline docs/LwM2M_Client/ESP32/ESP32-DevKitCBG96.md:22]: `coaps_uri://eu.iot.avsystem.cloud:5684` hehe
```suggestion
        - **Server URI** - the address and port of your {{ coiote_short_name }} installation, e.g. ``{{ coaps_uri }}``
```

File: docs/LwM2M_Client/ESP32/M5StickBG96.md [hunk 5]
```diff
@@ -19,7 +19,7 @@ Integrate your ESP32-based device with Quectel BG96 module to manage it via Coio
       ![BG96 UART example configuration](images/BG96_uart_config.png "BG96 UART example configuration")
     - In **Client options**, provide device credentials and Server URI:
         - **Endpoint name** - your device endpoint name
-        - **Server URI** - the address and port of your Coiote DM installation, e.g. ``coaps://eu.iot.avsystem.cloud:5684``
+        - **Server URI** - the address and port of your {{ coiote_short_name }} installation, e.g. ``coaps_uri://eu.iot.avsystem.cloud:5684``
```
Comment (JZimnol) [inline docs/LwM2M_Client/ESP32/M5StickBG96.md:22]: Same as above, change ``{{ coaps_uri }}``

File: main.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def define_env(env):
+
+    @env.macro
+    def console(description, txt):
+        # Check if image exists already and has the same hash, if so, return the image
+        hash = description + txt
+        hash = hashlib.md5(hash.encode('utf-8')).hexdigest()
+        file_path = 'images/console_gen_' + hash + '.png'
```
Comment (JZimnol) [inline main.py:14]: Maybe use `os.path.join`?

File: main.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        p = os.path.dirname(env.page.url).split("/")
+        p = p[:-1]
```
Comment (JZimnol) [inline main.py:16]:
```suggestion
        p = os.path.dirname(env.page.url).split("/")[:-1]
```
probably will work

File: main.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        relative_path = os.path.join(*p, file_path)
+        p = os.path.join("docs", *p, file_path)
+
+        documentation_image_path = '![' + description + \
+            '](' + file_path + " \"" + description + "\")"
```
Comment (JZimnol) [inline main.py:21]:
```suggestion
        documentation_image_path = '![' + description + '](' + file_path + " \"" + description + "\")"
```
For Python please use 120 characters per line.

File: main.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        if os.path.isfile(p):
+            return documentation_image_path
+
+        font_size = 16
+
+        mono_font = ImageFont.truetype(
+            "DejaVuSansMono.ttf", font_size, index=0)
```
Comment (JZimnol) [inline main.py:28]:
```suggestion
        mono_font = ImageFont.truetype("DejaVuSansMono.ttf", font_size, index=0)
```

File: main.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        img = Image.new('RGB', (10, 10), color=(0, 0, 0))
+        draw = ImageDraw.Draw(img)
+        (_, _, width, height) = draw.multiline_textbbox(
+            (0, 0), txt, font=mono_font, align="left")
```
Comment (JZimnol) [inline main.py:32]:
```suggestion
        (_, _, width, height) = draw.multiline_textbbox((0, 0), txt, font=mono_font, align="left")
```

**Review by JZimnol — COMMENTED** [review]
Aside my comments, what about that:

```
ag "Coiote" docs/

docs/cloud-integrations.md
3:With the Coiote IoT Device Management platform, you can integrate your LwM2M devices with Microsoft Azure, AWS and nRF Cloud.

docs/Cloud_integrations/Azure_IoT/Azure_IoT_Hub/Device_operations/Synchronize_devices_with_Azure_IoT_Hub.md
68:        The logs are only available in the previous version of Coiote. Select **Go to previous version** in the top-right corner.

docs/Cloud_integrations/Azure_IoT/Azure_IoT_Hub/Device_operations/Upgrading_firmware.md
21:    For firmware file recognition in Coiote, global identifiers are used. This means that it is recommended to name your firmware files using the format: yourdomainName + randomized value. If the same firmware file name is used again, then {{ coiote_short_name }} will be able to utilize the once downloaded resource without the need to download it again.

docs/Cloud_integrations/Azure_IoT/Azure_IoT_Hub/Perform_Read_Write.md
20:5. [A connected device](https://iotdevzone.avsystem.com/docs/Coiote_DM_Device_Onboarding/Quick_start/).

docs/Cloud_integrations/Azure_IoT/Azure_IoT_Hub/Perform_Execute.md
18:5. [A connected device](https://iotdevzone.avsystem.com/docs/Coiote_DM_Device_Onboarding/Quick_start/).

docs/Cloud_integrations/Azure_IoT/Azure_IoT_Hub/Set_an_Observation.md
20:5. [A connected device](https://iotdevzone.avsystem.com/docs/Coiote_DM_Device_Onboarding/Quick_start/).
96:1. In Coiote, go to **Device inventory --> Objects**.
118:3. In Coiote, you can now go to **Device inventory --> Objects** and see how **Value tracking** and **Attributes** boxes are now unchecked.

docs/Cloud_integrations/Azure_IoT/Azure_IoT_Central/Set_an_Observation.md
21:5. A connected device in {{ coiote_short_name }} that has been [exported to](https://iotdevzone.avsystem.com/docs/Azure_IoT_Integration_Guide/Azure_IoT_Central_integration/Device_operations/Exporting_devices_to_Azure_IoT_Central/) or [imported from](https://iotdevzone.avsystem.com/docs/Azure_IoT_Integration_Guide/Azure_IoT_Central_integration/Device_operations/Importing_devices_to_Coiote_DM/) Azure IoT Central.

docs/Cloud_integrations/AWS_IoT_Core/Overview.md
3:Explore the guide to learn how to integrate your IoT assets with the Coiote IoT Device Management platform.

docs/Cloud_integrations/AWS_IoT_Core/Configuring_AWS_integration.md
37:    - In the **Client Roles** fields, pick the **CoioteDM** client and **awsiottenant** role.
98:    - **coioteDMrestUsername** - username of the created CoioteDM account.
99:    - **coioteDMrestPassword** - password of the created CoioteDM account.
120:1. In **CoioteDM** remove all the devices from the **AWSiotCoreCertAuth** group.

docs/Demo_Projects/Implementing_LwM2M_objects_on_RaspberryPi.md
20:- A user with access to the Coiote IoT Device Management platform and appropriate permissions.
262:0. Refresh device state in Coiote and see if the Temperature object `/3303` has appeared in the Objects tab and values are reported for the Sensor Value `5700` resource:
496:0. Refresh device state in Coiote and see if the Multiple Axis Joystick object `/3345` has appeared in the Objects tab and values are reported for the Digital Input State `5500` and Digital Input Counter `5501` resources:

docs/Demo_Projects/Edge_Impulse_demo.md
92:## Connect the Thingy:91 to Coiote IoT Device Management
113:## Add the Pattern Detection Object to Coiote
210:For more information, see [how to set observations](/Coiote_IoT_DM/Device_Center/#set-observation)

docs/Demo_Projects/Tracking_tutorial.md
26:This tutorial uses the **Thingy:91** prototyping platform in combination with the **Coiote IoT Device Management** platform to build a cellular-connected tracking application. The integration with **nRF Cloud Locator** enables cell-based localization and optimizes the usage of the onboard GNSS. The location data, in combination with additional telemetry data is sent to **Azure IoT Hub** and visualized using **Microsoft Power BI**.
32:## Part 1 - Connect the Thingy:91 to Coiote using the LwM2M Anjay client
106:## Connect the Thingy:91 to Coiote IoT Device Management
157:    * To update the **endpoint name**, enter the endpoint name you created in Coiote:
163:    * To update the **Pre-Shared Key**, enter the key you created in Coiote:
184:If the connection to nRF Cloud Locator was successful, you will see the device location as a widget in the Coiote Device Center.
189:## Part 3 - Connect Coiote to Microsoft Azure
195:1. In Coiote, click **Integrations** from the left-side menu and select **Hyperscaler Integration Center**.
262:In Coiote, visit your **Device inventory**. Find the device you want to connect to Azure, click the three dots icon and select **Connect to Azure**.

docs/quickstart.md
5:This Quick Start walks you through the process of building your first LwM2M application using AVSystem’s LwM2M Client **Anjay** and AVSystem’s LwM2M Server **Coiote IoT Device Management**.
13:Get started by <a href="{{ coiote_site_link }}/" target="_blank">**signing up**</a> to Coiote IoT Device Management. You can sign up for a developer account, which allows you to connect up to 10 devices free of charge.
17:![Coiote Device Inventory](img/Empty-Device-Inventory.png "Empty device inventory at {{ coiote_short_name }}")
126:Copy the command you find in Coiote.
134:If the demo application started successfully, a device is shown in the Coiote platform.
136:![Coiote dashboard](img/coiote-dashboard.png)
185:Dive into Coiote IoT Device Management and explore its numerous functionalities. From device onboarding using the [Bootstrap Server](/Coiote_IoT_DM/Onboarding/bootstrap/) to [Data Visualization](/Coiote_IoT_DM/Quick_Start/Visualize_device_data/) and running automated [Device Tests](/Coiote_IoT_DM/Device_tests/Overview/).
187:[{{ coiote_short_name }} docs](/Coiote_IoT_DM/introduction/){: .md-button .md-button--big }

docs/index.md
14:In the DevZone you will find documentation related to our scalable LwM2M Server **Coiote IoT Device Management**, as well as our LwM2M Client **Anjay**.
21:The guide walks you through the process of building your first LwM2M application using AVSystem’s LwM2M Client **Anjay** and AVSystem’s LwM2M Server **Coiote IoT Device Management**.
37:Dive into Coiote IoT Device Management, the scalable LwM2M server and explore its numerous features. From device onboarding using the [Bootstrapping Server](Coiote_IoT_DM/User_Interface/Device_onboarding/) to [Data Visualization](Coiote_IoT_DM/Quick_Start/Visualize_device_data/) and running automated [Device Tests](Coiote_IoT_DM/Device_tests/Overview/).
39:[{{ coiote_short_name }} docs](Coiote_IoT_DM/introduction/){: .md-button .md-button--big }
44:With the Coiote IoT Device Management platform, you can easily integrate your LwM2M data with **Azure IoT Hub**, **Azure IoT Central**, **Azure DPS**, **AWS IoT Core** and **nRF Cloud Location Services**.

docs/Coiote_IoT_DM/multi_firmware_update.md
55:1. In the Coiote IoT Device Management platform, go to [**Device Inventory**]({{ coiote_site_link }}/ui/device/inventory).

docs/Coiote_IoT_DM/firmware_update.md
58:1. In the Coiote IoT Device Management platform, go to [**Device Inventory**]({{ coiote_site_link }}/ui/device/inventory).

docs/Coiote_IoT_DM/Device_tests/Overview.md
3:Device tests - also known as hardware in the loop tests - is a solution to test the performance of LwM2M devices and is offered as part of the **Coiote IoT Device Management** platform. Device tests are an essential tool in the development process of device firmware as they can be used for:

docs/Coiote_IoT_DM/introduction.md
6:# Coiote IoT Device Management
8:Coiote IoT Device Management Platform allows you to manage your LwM2M devices throughout their entire lifecycle. Coiote supports device onboarding, data management, data visualization, automated tests, firmware updates, monitoring & alerting and cloud integrations.
13:    Get started by <a href="{{ coiote_site_link }}/" target="_blank">**signing up**</a> to Coiote IoT Device Management. You can sign up for a developer account which allows you to connect up to 10 devices free of charge.
15:![Coiote login page](images/login-page-new.png "Coiote login page")
27:→ **Learn more about the [**Device center**](/Coiote_IoT_DM/Device_Center/).**
35:→ **Learn more about [**Data visualization**](/Coiote_IoT_DM/Visualize_data/).**
46:    Features marked with an "*****" are only included in the premium version of Coiote.
63:→ **Learn more about [**Device tests**](/Coiote_IoT_DM/Device_tests/Overview/).**

docs/academy/module5.md
24:*<p style="text-align: center;">Updating a device’s firmware using Coiote IoT Device Management</p>*
80:* [DevZone Firmware update documentation](https://iotdevzone.avsystem.com/docs/Coiote_IoT_DM/firmware_update/)

docs/academy/exercise3b.md
387:If all went well and logs show **registration successfully updated**, you can go to Coiote IoT Device Management where the Registration status should show **Registered**. In the **Data model** view, you should now see the Temperature Object containing all the created Resources.

docs/academy/exercise2b.md
15:For LwM2M Servers like [{{ coiote_short_name }}](https://www.avsystem.com/coiote-iot-device-management-platform/), server-side configuration needs to be updated to enable LwM2M Clients to connect using PSK. The simplest solution is to edit the connection parameters on the Coiote IoT Device Management configuration page. Follow the next steps to change the parameters and set them to Pre-Shared Key mode.
172:In Coiote IoT Device Management, the Registration status should show **Registered** and the **Security mode** in the bottom-left corner panel should now include the tag: `Pre-Shared Key`.

docs/academy/exercise4b.md
277:Now, open Coiote IoT Device Management and validate if the Resources are updated after each configured time interval.

docs/academy/module4.md
51:*<p style="text-align: center;">Set Observation in Coiote IoT Device Management Platform</p>*

docs/academy/exercise3a.md
432:If all went well and logs show **registration successfully updated**, you can go to Coiote IoT Device Management where the Registration status should show **Registered**. Go to the **Data Model** and check if you can see the Time Object in the **Defined objects** section.

docs/academy/exercise1.md
87:Before connecting your device to the cloud, log in to Coiote IoT Device Management: <a href="{{ coiote_site_link }}" target="_blank">**eu.iot.avsystem.cloud**</a>.
175:Did you manage to connect the Raspberry Pi Pico W to the Coiote IoT Device Management server?

docs/academy/exercise4a.md
19:Let’s get started by connecting our Raspberry Pi Pico W to the computer via a USB cable. Open a serial communication program to watch the device logs and go to the Coiote IoT Device Management to check if your device has connected successfully. If so, you are ready to perform your first Observe/Read operations on your LwM2M Device (a.k.a. The LwM2M Client).

docs/academy/index.md
76:Coiote IoT Device Management is a platform built around the LwM2M standard. It acts as a LwM2M Server and is designed as a scalable platform to manage large numbers of devices. The platform comes with tools to monitor groups of devices, manage device configurations, and perform firmware updates. In addition, it supports cloud integrations with platforms such as AWS IoT Core or Azure IoT Hub.
98:Find [more information](https://iotdevzone.avsystem.com/docs/) about getting started with Coiote IoT Device Managements.

docs/academy/exercise5.md
5:Your graduation project involves building an application that utilizes the LwM2M standard and the Coiote IoT Device Management platform. Whether you choose the Raspberry Pi Pico W or any other IoT device, your task is to effectively gather sensor data and establish a cloud connection.

docs/LwM2M_Client/Nordic/Thingy91.md
172:3. Go to the Coiote DM to check if your device is connected.

docs/LwM2M_Client/Nordic/Device_Provisioning.md
28:    - Edit `lwm2m_server.json` modify `domain` entry to reflect your domain in Coiote server. This file is needed if you wish the script to automatically add the new device to {{ coiote_short_name }}.
92:    You may need to remove the device from Coiote if you finished the steps in previous section and the device is already registered. Coiote will not allow registration of the device with the same name.
101:    - Edit `lwm2m_server.json` modify `domain` entry to reflect your domain in Coiote server.
114:2. Getting Coiote Access Token
139:4. Connecting device to Coiote
149:    If everyting went well you should see your new certificate and the device should be ready to connect to Coiote.

docs/LwM2M_Client/OpenThread/Adding_device.md
11:- A user with access to the [**Coiote IoT Device Management**]({{ coiote_site_link }}/ui/device/inventory) platform and appropriate permissions.

docs/LwM2M_Client/ESP32/ESP32-DevKitCBG96.md
9:- A user with access to the [Coiote IoT Device Management platform]({{ coiote_site_link }}).

docs/LwM2M_Client/ESP32/M5Stick.md
9:- A user with access to the Coiote IoT Device Management platform.

docs/LwM2M_Client/ESP32/M5StickBG96.md
9:- A user with access to the [Coiote IoT Device Management platform]({{ coiote_site_link }}).

docs/LwM2M_Client/STMicroelectronics/B-U585I-IOT02A-BG96.md
9:- A user with access to the Coiote IoT Device Management platform.

docs/LwM2M_Client/STMicroelectronics/B-L475E-IOT01A.md
10:- A user with access to the Coiote IoT Device Management platform and appropriate permissions.

docs/LwM2M_Client/RaspberryPi/RaspberryPi-Pico-W.md
8:- A user with access to the Coiote IoT Device Management platform and appropriate permissions.
```
There is a lot of `Coiote IoT Device Management` etc. that can be changed to macros. Is there any reason why you didn't replace it?

File: main.py (hunk 8 above)
Comment (Mierunski) [reply main.py:20]: This is autopep8 default, not changing this one

**Review by JZimnol — APPROVED** [review]
@Mierunski looks ~good~ better. Approved.
